egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py

```diff
@@ -1851,3 +1853,224 @@ def modified_beam_search_ngram_rescoring(
+
+        logits = logits.squeeze(1).squeeze(1)  # (num_hyps, vocab_size)
+
+        log_probs = logits.log_softmax(dim=-1)  # (num_hyps, vocab_size)
+
+        log_probs.add_(ys_log_probs)
+
+        vocab_size = log_probs.size(-1)
+
+        log_probs = log_probs.reshape(-1)
+
+        row_splits = hyps_shape.row_splits(1) * vocab_size
+        log_probs_shape = k2.ragged.create_ragged_shape2(
+            row_splits=row_splits, cached_tot_size=log_probs.numel()
+        )
+        ragged_log_probs = k2.RaggedTensor(
+            shape=log_probs_shape, value=log_probs
+        )
+
+        # for all hyps with a non-blank new token, score it
+        token_list = []
+        hs = []
+        cs = []
+        for i in range(batch_size):
+            topk_log_probs, topk_indexes = ragged_log_probs[i].topk(beam)
+
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore")
+                topk_hyp_indexes = (topk_indexes // vocab_size).tolist()
+                topk_token_indexes = (topk_indexes % vocab_size).tolist()
+            for k in range(len(topk_hyp_indexes)):
+                hyp_idx = topk_hyp_indexes[k]
+                hyp = A[i][hyp_idx]
+
+                new_token = topk_token_indexes[k]
+                if new_token not in (blank_id, unk_id):
+
+                    assert new_token != 0, new_token
+                    token_list.append([new_token])
+                    hs.append(hyp.state[0])
+                    cs.append(hyp.state[1])
+        # forward RNNLM to get new states and scores
+        if len(token_list) != 0:
+            tokens_to_score = (
+                torch.tensor(token_list)
+                .to(torch.int64)
+                .to(device)
+                .reshape(-1, 1)
+            )
+
+            hs = torch.cat(hs, dim=1).to(device)
+            cs = torch.cat(cs, dim=1).to(device)
+            scores, lm_states = rnnlm.score_token(tokens_to_score, (hs, cs))
+
+        count = 0  # index, used to locate score and lm states
+        for i in range(batch_size):
+            topk_log_probs, topk_indexes = ragged_log_probs[i].topk(beam)
+
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore")
+                topk_hyp_indexes = (topk_indexes // vocab_size).tolist()
+                topk_token_indexes = (topk_indexes % vocab_size).tolist()
+
+            for k in range(len(topk_hyp_indexes)):
+                hyp_idx = topk_hyp_indexes[k]
+                hyp = A[i][hyp_idx]
+
+                ys = hyp.ys[:]
+
+                lm_score = hyp.lm_score
+                state = hyp.state
+
+                hyp_log_prob = topk_log_probs[k]  # get score of current hyp
+                new_token = topk_token_indexes[k]
+                if new_token not in (blank_id, unk_id):
+
+                    ys.append(new_token)
+                    hyp_log_prob += (
+                        lm_score[new_token] * lm_scale
```

Comment:
If I understand correctly, the ```lm_scale``` is for scaling the previous lm_score (not the one from the RNN-LM), is that correct ?

Reply:
It's not a typo. Each time when we create a new hypothesis, we pass a variable `lm_score` to it, which is the RNNLM score for the next token given the current transcription, i.e `ys`. Therefore, when we encounter a non-blank token, we just need to add `hyp.lm_score * lm_scale` to `hyp.log_prob`

Reply:
Ah got it. Another question though, is there a reason why when rescoring with an N-gram, the lm_score is removed :
```
# We only keep AM scores in new_hyp.log_prob
                new_log_prob = (
                    topk_log_probs[k] - hyp.state_cost.lm_score * lm_scale
                )
```

but not in this case ?

Reply:
Well, we also don't know for sure. But I am investigating this.

Also, could you please remind me which text corpora was the RNNLM in this repo (https://huggingface.co/ezerhouni/icefall-librispeech-rnn-lm/tree/main) trained on?

Reply:
I trained on librispeech train data : 
https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/prepare.sh#L283

Reply:
> Ah got it. Another question though, is there a reason why when rescoring with an N-gram, the lm_score is removed :
> 
> ```
> # We only keep AM scores in new_hyp.log_prob
>                 new_log_prob = (
>                     topk_log_probs[k] - hyp.state_cost.lm_score * lm_scale
>                 )
> ```
> 
> but not in this case ?

I found the reason. The n-gram LM returns a cumulative score given the transcription whereas the RNNLM just gives the score for the next token.